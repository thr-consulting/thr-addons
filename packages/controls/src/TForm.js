// @flow

import React, {Component} from 'react';
import isEmpty from 'lodash/isEmpty';
import {Formik} from 'formik';
import {Message} from 'semantic-ui-react';

type Props = {
	/** Render prop just like Formik's with additional arguments passed */
	render: any,
	/** The number of fields in the form. TForm can't automatically get this number. */
	numFields: number,
	/** Used to pass in any GraphQL errors. */
	errors?: any,
	/** Called when the form submits. */
	onSubmit: data => void,
};

/**
 * Extends Formik to provide Semantic UI error and warning messages and field errors.
 * @class
 * @property {function} render - Render prop just like Formik's with additional arguments passed
 * @property {number} numFields - The number of fields in the form. TForm can't automatically get this number.
 * @property errors - Used to pass in any GraphQL errors.
 * @property {function} onSubmit - Called when the form submits.
 */
export default class TForm extends Component<Props> {
	static displayName = 'TForm';

	props: Props;

	renderForm = args => {
		const {errors: warnings, touched} = args;
		const {errors, numFields, render} = this.props;

		return render({
			renderErrors() {
				if (!errors) return null;
				const {graphQLErrors} = errors;
				return (
					<Message error>
						<Message.Header>Error</Message.Header>
						{graphQLErrors.map(err => (
							<p key={err.message}>{err.message}</p>
						))}
					</Message>
				);
			},
			renderWarnings() {
				if (isEmpty(warnings)) return null;
				return (
					<Message warning>
						<Message.Header>Please check the following fields:</Message.Header>
						<ul>
							{Object.keys(warnings).map(warning => (
								<li key={warning}>{warnings[warning]}</li>
							))}
						</ul>
					</Message>
				);
			},
			hasErrors() {
				return !!errors;
			},
			hasWarnings() {
				return Object.keys(touched).length === numFields && !isEmpty(warnings);
			},
			fieldError(fieldName) {
				return !!(touched[fieldName] && warnings[fieldName]);
			},
			...args,
		});
	};

	render() {
		const {render, numFields, ...rest} = this.props;
		return (
			<Formik render={this.renderForm} {...rest}/>
		);
	}
}
