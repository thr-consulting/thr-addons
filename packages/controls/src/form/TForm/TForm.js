// @flow

import React, {Component} from 'react';
import isEmpty from 'lodash/isEmpty';
import isString from 'lodash/isString';
import {Formik} from 'formik';
import {Message} from 'semantic-ui-react';

type Props = {
	/** Render prop just like Formik's with additional arguments passed */
	render: any,
	/** The number of fields in the form. TForm can't automatically get this number. */
	numFields: number,
	/** Used to pass in any GraphQL errors. */
	errors?: any,
	/** Called when the form submits. */
	onSubmit: data => void,
	/** Called when the TForm is mounted so you can get access to the submitForm function. */
	getSubmitFn: fn => void,
};

/**
 * Extends Formik to provide Semantic UI error and warning messages and field errors.
 * @class
 * @property {function} render - Render prop just like Formik's with additional arguments passed
 * @property {number} numFields - The number of fields in the form. TForm can't automatically get this number.
 * @property errors - Used to pass in any GraphQL errors.
 * @property {function} onSubmit - Called when the form submits.
 * @property {function} getSubmitFn - Called when the TForm is mounted so you can get access to the submitForm function.
 */
export default class TForm extends Component<Props> {
	static displayName = 'TForm';

	componentDidMount() {
		if (this.props.getSubmitFn) this.props.getSubmitFn(this._submitForm);
	}


	renderForm = args => {
		const {errors: warnings, touched, handleChange, setFieldValue, submitForm, ...rest} = args;
		const {errors, numFields, render} = this.props;

		this._submitForm = submitForm;

		return render({
			renderErrors() {
				if (!errors) return null;
				const {graphQLErrors} = errors;
				return (
					<Message error>
						<Message.Header>Error</Message.Header>
						{graphQLErrors.map(err => (
							<p key={err.message}>{err.message}</p>
						))}
					</Message>
				);
			},
			renderWarnings() {
				if (isEmpty(warnings)) return null;
				return (
					<Message warning>
						<Message.Header>Some fields are not complete:</Message.Header>
						<ul>
							{Object.keys(warnings).map(warning => (
								<li key={warning}>{warnings[warning]}</li>
							))}
						</ul>
					</Message>
				);
			},
			hasErrors() {
				return !!errors;
			},
			hasWarnings() {
				return Object.keys(touched).length === numFields && !isEmpty(warnings);
			},
			fieldError(fieldName) {
				return !!(touched[fieldName] && warnings[fieldName]);
			},
			handleChange(evOrName) {
				if (evOrName.nativeEvent) return handleChange(evOrName);
				if (isString(evOrName)) {
					return val => {
						setFieldValue(evOrName, val);
					};
				}
				throw new Error('TForm expects handleChange to receive a SyntheticEvent or a string value of the field name');
			},
			errors: warnings,
			touched,
			setFieldValue,
			submitForm,
			...rest,
		});
	};

	render() {
		const {render, numFields, ...rest} = this.props;
		return (
			<Formik render={this.renderForm} {...rest}/>
		);
	}
}
