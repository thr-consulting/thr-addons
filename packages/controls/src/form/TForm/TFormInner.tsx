import debug from 'debug';
import React, {ChangeEvent, useState, useEffect} from 'react';
import {FormikProps, FormikErrors, FormikTouched} from 'formik';
import {Message, Segment} from 'semantic-ui-react';
import flatten from 'flat';
import get from 'lodash/get';
import uniq from 'lodash/uniq';

const d = debug('thx.controls.TFormInner');

const isFunction = (obj: any): obj is Function => typeof obj === 'function';
const isEmptyChildren = (children: any): boolean => React.Children.count(children) === 0;

export interface TFormChildrenProps<Values> extends FormikProps<Values> {
	renderWarnings: () => JSX.Element; // Returns a Message component displaying warnings and errors.
	hasErrors: boolean; // True if errors have been passed from parent but has not been marked as cleared
	hasWarnings: boolean; // True if form validation warnings occur
	fieldError: (fieldName: keyof Values) => boolean;
	loading?: boolean; // We just pass this variable from parent to child
	errorMarkedCleared: boolean; // Errors have been marked as cleared
	submitDisabled: boolean; // Semantic UI Convenience variable
	formError: boolean; // Semantic UI Convenience variable
}

// I don't want install the apollo package here, so this is the general type for error:
interface ApolloError {
	message?: string;
	graphQLErrors?: {
		message?: string;
	}[];
}

export interface TFormProps<Values> {
	children?: ((props: TFormChildrenProps<Values>) => React.ReactNode) | React.ReactNode;
	numFields?: number;
	loading?: boolean;
	error?: ApolloError; // Graphql errors
	onValidate?: (isValid: boolean) => void;
	getSubmitFn?: (submitFormFn: () => Promise<void> | Promise<any>) => void;
	onChange?: (values: Values) => void;
}

export interface TFormInnerProps<Values> {
	formikProps: FormikProps<Values>;
	tFormProps: TFormProps<Values>;
	children?: ((props: TFormChildrenProps<Values>) => React.ReactNode) | React.ReactNode;
}

export function TFormInner<Values>(props: TFormInnerProps<Values>): JSX.Element | null {
	const {children, formikProps, tFormProps} = props;

	useEffect(() => {
		if (tFormProps.onValidate) tFormProps.onValidate(formikProps.isValid);
	}, [formikProps.isValid, tFormProps.onValidate]);

	// Get submit function
	useEffect(() => {
		if (tFormProps.getSubmitFn) tFormProps.getSubmitFn(formikProps.submitForm);
	}, [formikProps.submitForm, tFormProps.getSubmitFn]);

	// On change event (if this ends up changing values it will cycle)
	useEffect(() => {
		if (tFormProps.onChange) tFormProps.onChange(formikProps.values);
	}, [formikProps.values]);

	// Mark the error cleared if the form has been edited after an error
	const [errorCleared, setErrorCleared] = useState(false);
	useEffect(() => {
		if (tFormProps.error && tFormProps.error.message && !formikProps.isSubmitting) {
			setErrorCleared(true);
		}
	}, [formikProps.values, tFormProps.error]);

	// Flatten the warnings (formik errors) and touch objects.
	const flatWarnings = flatten<FormikErrors<Values>, {[key: string]: string}>(formikProps.errors);
	const flatTouched = flatten<FormikTouched<Values>, {[key: string]: string}>(formikProps.touched);

	// Create an array of strings that are the warnings (formik errors). If the submit count is greater than 0
	// then it will show all warnings. Otherwise only touched field warnings will be added to this array.
	const warnings: string[] = uniq(
		Object.keys(flatWarnings).reduce<string[]>((memo, value) => {
			const isTouched = flatTouched[value];
			if (isTouched || formikProps.submitCount > 0) {
				return [...memo, get(flatWarnings, value)];
			}
			return memo;
		}, []),
	);

	let errors: string[] = [];
	let errorHeader = '';
	if (tFormProps.error && tFormProps.error.message) {
		if (tFormProps.error.graphQLErrors) {
			errors = tFormProps.error.graphQLErrors.reduce((memo, v) => {
				if (v.message) return [...memo, v.message];
				return memo;
			}, [] as string[]);
		} else {
			const errorMessage = tFormProps.error.message.slice(tFormProps.error.message.indexOf(': ') + 1);
			errors.push(errorMessage);
		}
		errorHeader = tFormProps.error.message.slice(0, tFormProps.error.message.indexOf(': '));
	}
	errors = uniq(errors);

	const hasWarnings = !(formikProps.isValid || warnings.length === 0);
	const hasErrors = errors.length > 0;

	const bag: TFormChildrenProps<Values> = {
		...formikProps,
		renderWarnings() {
			if (!hasWarnings && !hasErrors) return null;
			return (
				<Message warning={hasWarnings} error={hasErrors && !hasWarnings}>
					<Message.Header>{hasWarnings ? 'Some fields are not complete:' : `${errorHeader}:`}</Message.Header>
					{/* Put it in a segment to make sure the errorMessage isn't too big when there are a lot of errors */}
					<Segment style={{overflow: 'auto', maxHeight: 100}}>
						{(hasWarnings ? warnings : errors).map(msg => (
							<div key={msg}>{msg}</div>
						))}
					</Segment>
				</Message>
			);
		},
		fieldError(fieldName) {
			return (formikProps.touched[fieldName] || formikProps.submitCount > 0) && !!formikProps.errors[fieldName];
		},
		handleChange(ev: ChangeEvent | string) {
			if (typeof ev === 'string') {
				d(ev);
				return (val: any) => {
					d(ev, val);
					formikProps.setFieldValue(ev, val);
				};
			}
			if (ev.nativeEvent) return formikProps.handleChange(ev);
			throw new Error('TForm expects handleChange to receive a SyntheticEvent or a string value of the field name');
		},
		handleSubmit(e?: React.FormEvent<HTMLFormElement>) {
			setErrorCleared(false);
			formikProps.handleSubmit(e);
		},
		hasErrors: hasErrors && !errorCleared,
		hasWarnings,
		isValid: !hasWarnings,
		loading: tFormProps.loading,
		submitDisabled: hasWarnings || formikProps.isSubmitting,
		formError: hasErrors && !hasWarnings && !errorCleared && !formikProps.isSubmitting, // Convenience
		errorMarkedCleared: errorCleared,
	};

	if (children) {
		if (isFunction(children)) {
			return (children as (bag: TFormChildrenProps<Values>) => JSX.Element)(bag);
		}
		if (!isEmptyChildren(children)) {
			// @ts-ignore
			return React.Children.only(children);
		}
		return null;
	}
	return null;
}
