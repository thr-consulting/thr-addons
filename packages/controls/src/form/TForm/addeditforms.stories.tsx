import {useArgs} from '@storybook/client-api';
import type {Meta} from '@storybook/react';
import debug from 'debug';
import {Form, Input} from 'semantic-ui-react';
import {number, object, string} from 'yup';
import type {TFormConfig, TFormProps} from './types';
import {useTForm} from './useTForm';

const d = debug('thx.controls.form.TForm.addeditforms.stories');

export default {
	title: 'Form/Add and Edit Forms',
	argTypes: {
		onSubmit: {type: 'function'},
		onChange: {type: 'function'},
		onValidate: {type: 'function'},
	},
} as Meta;

interface LockedEditForm {
	id: string;
	version: number;
}

function lockedEditFormValidation(a: any) {
	return a.clone().shape({
		id: string().required(),
		version: number().required(),
	});
}

/* ***************************************************************************** */

const addressFormValidation = object().shape({
	line1: string().required(),
	line2: string().notRequired(),
});
const addressEditFormValidation = lockedEditFormValidation(addressFormValidation);

interface AddressFormType {
	line1: string;
	line2?: string;
}

type AddressEditFormType = AddressFormType & LockedEditForm;

function AddressForm<T extends AddressFormType>(props: TFormProps<T>) {
	const {handleSubmit, formError, hasWarnings, values, handleChange, handleBlur, submitDisabled, isSubmitting} = props;

	return (
		<Form onSubmit={handleSubmit} error={formError} warning={hasWarnings}>
			<Form.Field width={6}>
				<label>Line 1</label>
				<Input name="line1" value={values.line1} onChange={handleChange} onBlur={handleBlur} />
			</Form.Field>
			<Form.Field width={6}>
				<label>Line 2</label>
				<Input name="line2" value={values.line2} onChange={handleChange} onBlur={handleBlur} />
			</Form.Field>
			<Form.Button disabled={submitDisabled} type="submit" loading={isSubmitting}>
				Submit
			</Form.Button>
		</Form>
	);
}

export function AddressCreateForm(args: {onSubmit: (value: any) => any}) {
	const [, updateArgs] = useArgs();

	const tform = useTForm<AddressFormType>({
		initialValues: {line1: '', line2: ''},
		validationSchema: addressFormValidation,
		onSubmit(value) {
			args.onSubmit && args.onSubmit(addressFormValidation.validateSync(value, {stripUnknown: true}));
		},
	});

	return <AddressForm {...tform} />;
}
AddressCreateForm.args = {
	enableReinitialize: false,
	loading: false,
};

export function AddressEditForm(args: TFormConfig<AddressEditFormType>) {
	const [, updateArgs] = useArgs();

	const tform = useTForm<AddressEditFormType>({
		...args,
		initialValues: {line1: '', line2: ''},
		validationSchema: addressEditFormValidation,
		async onSubmit(value, helpers) {
			await args.onSubmit(addressEditFormValidation.validateSync(value, {stripUnknown: true}), helpers);
		},
	});

	return <AddressForm {...tform} />;
}
AddressEditForm.args = {
	enableReinitialize: false,
	loading: false,
};
