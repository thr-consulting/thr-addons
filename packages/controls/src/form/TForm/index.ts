export {default} from './TForm';
export {TFormProps, TFormChildrenProps, TFormInner, TFormInnerProps} from './TFormInner';
