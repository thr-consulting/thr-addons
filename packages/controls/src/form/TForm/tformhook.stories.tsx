import debug from 'debug';
import React, {useRef, useState} from 'react';
import {Form, Container, Button, ButtonGroup, Segment} from 'semantic-ui-react';
import {object, string} from 'yup';
import {useTForm} from './useTForm';

const d = debug('thx.controls.useTForm.stories');

export default {title: 'useTForm'};

// Yup Validation
const formValidation = object().shape({
	text: string().required('Text is required'),
	nested: object().shape({
		text: string().required('Nested text is required'),
	}),
	plainInput: string(),
	customInput: string(),
});

interface FormValidation {
	text: string;
	nested: {
		text: string;
	};
	plainInput: string;
}

interface ApolloError {
	message?: string;
	graphQLErrors?: {
		message?: string;
	}[];
}

const sampleGraphqlError: (ApolloError | undefined)[] = [
	undefined,
	{message: 'Sample Graphql Error 1'},
	{
		message: 'Sample Graphql Error 2',
		graphQLErrors: [{message: 'Graphql Sub Error 2-A'}, {message: 'Graphql Sub Error 2-B'}],
	},
];

export const Main = () => {
	const [vars, setVars] = useState({
		graphqlError: 0,
	});
	const [graphqlError, setGraphqlError] = useState<ApolloError>();
	const formSubmitFn = useRef<() => Promise<void>>();

	const {
		values,
		handleChange,
		handleBlur,
		handleSubmit,
		renderWarnings,
		formError,
		hasWarnings,
		submitDisabled,
		fieldError,
		isSubmitting,
	} = useTForm<FormValidation>({
		initialValues: {
			text: '',
			nested: {
				text: '',
			},
			plainInput: '',
		},
		validationSchema: formValidation,
		error: graphqlError,
		getSubmitFn: sub => (formSubmitFn.current = sub),
		onSubmit: data =>
			new Promise((resolve, reject) => {
				setTimeout(() => {
					if (vars.graphqlError === 0) {
						d('Success!', data);
						resolve(data);
					} else {
						setGraphqlError(sampleGraphqlError[vars.graphqlError]);
						reject();
					}
				}, 1000);
			}),
	});

	return (
		<Container>
			<Form error={formError} warning={hasWarnings} onSubmit={handleSubmit}>
				<Form.Field width={6} error={fieldError('text')}>
					<label>Enter some text</label>
					<input name="text" value={values.text} onChange={handleChange} onBlur={handleBlur} />
				</Form.Field>
				<Form.Field width={6} error={fieldError('nested')}>
					<label>Enter some nested text</label>
					<input name="nested.text" value={values.nested.text} onChange={handleChange} onBlur={handleBlur} />
				</Form.Field>
				<Form.Field error={fieldError('plainInput')}>
					<label>Plain input</label>
					<input name="plainInput" value={values.plainInput} onChange={handleChange} onBlur={handleBlur} />
				</Form.Field>
				<Form.Button disabled={submitDisabled} type="submit" loading={isSubmitting}>
					Submit
				</Form.Button>
				{renderWarnings()}
			</Form>
			<Segment>
				<ButtonGroup vertical>
					<Button
						color={vars.graphqlError === 0 ? 'green' : 'blue'}
						onClick={() => {
							setVars({graphqlError: 0});
							setGraphqlError(undefined);
						}}
					>
						Clear GraphQL Error
					</Button>
					<Button
						color={vars.graphqlError === 1 ? 'green' : 'blue'}
						onClick={() => {
							setVars({graphqlError: 1});
							setGraphqlError(undefined);
						}}
					>
						Simulate GraphQL Error 1
					</Button>
					<Button
						color={vars.graphqlError === 2 ? 'green' : 'blue'}
						onClick={() => {
							setVars({graphqlError: 2});
							setGraphqlError(undefined);
						}}
					>
						Simulate GraphQL Error 2
					</Button>
					<Button color="orange" onClick={() => formSubmitFn.current && formSubmitFn.current()}>
						Outside Submit
					</Button>
				</ButtonGroup>
			</Segment>
		</Container>
	);
};
