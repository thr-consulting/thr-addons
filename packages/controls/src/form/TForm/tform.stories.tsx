import {Button, Group, TextInput} from '@mantine/core';
/* eslint-disable react-hooks/rules-of-hooks */
import {useArgs} from '@storybook/client-api';
import type {Meta} from '@storybook/react';
import debug from 'debug';
import {useState} from 'react';
import {object, string} from 'yup';
import {storyDecorator} from '../../storyDecorator';
import {TForm} from './TForm';
import type {TFormProps} from './types';

const d = debug('thx.controls.form.TForm.tform.stories');

export default {
	title: 'Form/TForm',
	argTypes: {
		onSubmit: {type: 'function'},
		onChange: {type: 'function'},
		onValidate: {type: 'function'},
	},
	decorators: [
		storyDecorator,
		Story => (
			<div style={{width: 400, position: 'relative'}}>
				<Story />
			</div>
		),
	],
} as Meta;

// Yup Validation
const formValidation = object().shape({
	text: string().required('Text is required'),
	nested: object().shape({
		text: string().required('Nested text is required'),
	}),
	plainInput: string(),
});

interface FormType {
	text: string;
	nested: {
		text: string;
	};
	plainInput: string;
}

function MyForm(props: TFormProps<FormType>) {
	const {values, handleChange, handleSubmit, handleBlur, submitDisabled, isSubmitting, renderWarnings} = props;

	// error={formError} warning={hasWarnings}
	return (
		<form onSubmit={handleSubmit}>
			<TextInput name="text" label="Enter some text" value={values.text} onChange={handleChange} onBlur={handleBlur} />
			<TextInput name="nested.text" label="Enter some nested text" value={values.nested.text} onChange={handleChange} onBlur={handleBlur} />
			<TextInput name="plainInput" label="Plain Input" value={values.plainInput} onChange={handleChange} onBlur={handleBlur} />
			<Group position="right" mt="md">
				<Button disabled={submitDisabled} type="submit" loading={isSubmitting}>
					Submit
				</Button>
			</Group>
			{renderWarnings()}
		</form>
	);
}

export function Main(args: any) {
	const [, updateArgs] = useArgs();

	return (
		<TForm<FormType>
			{...args}
			initialValues={{text: '', nested: {text: ''}, plainInput: ''}}
			validationSchema={formValidation}
			onSubmit={async (data, helpers) => {
				args.onSubmit(data, helpers);
			}}
		>
			{props => <MyForm {...props} />}
		</TForm>
	);
}
Main.args = {
	enableReinitialize: false,
	loading: false,
};

/* **************************************************************************** */

interface ApolloError {
	message?: string;
	graphQLErrors?: {
		message?: string;
	}[];
}

const sampleGraphqlError: (ApolloError | undefined)[] = [
	undefined,
	{message: 'Sample Graphql Error 1'},
	{
		message: 'Sample Graphql Error 2',
		graphQLErrors: [{message: 'Graphql Sub Error 2-A'}, {message: 'Graphql Sub Error 2-B'}],
	},
];

export function WithError(args: any) {
	// const [, updateArgs] = useArgs();
	const [graphqlError, setGraphqlError] = useState(0);

	return (
		<>
			<div style={{width: 400, position: 'relative'}}>
				<TForm<FormType>
					{...args}
					error={sampleGraphqlError[graphqlError]}
					initialValues={{text: '', nested: {text: ''}, plainInput: ''}}
					validationSchema={formValidation}
					onSubmit={data => {
						return new Promise((resolve, reject) => {
							setTimeout(() => {
								if (graphqlError === 0) {
									args.onSubmit(data);
									resolve(data);
								} else {
									// updateArgs({error: sampleGraphqlError[graphqlError]});
									reject();
								}
							}, 1000);
						});
					}}
				>
					{props => <MyForm {...props} />}
				</TForm>
			</div>
			<Group position="left" mt="md">
				<Button
					color={graphqlError === 0 ? 'green' : 'blue'}
					onClick={() => {
						setGraphqlError(0);
						// updateArgs({error: undefined});
					}}
				>
					Clear GraphQL Error
				</Button>
				<Button
					color={graphqlError === 1 ? 'green' : 'blue'}
					onClick={() => {
						setGraphqlError(1);
						// updateArgs({error: undefined});
					}}
				>
					Simulate GraphQL Error 1
				</Button>
				<Button
					color={graphqlError === 2 ? 'green' : 'blue'}
					onClick={() => {
						setGraphqlError(2);
						// updateArgs({error: undefined});
					}}
				>
					Simulate GraphQL Error 2
				</Button>
			</Group>
		</>
	);
}
WithError.args = {
	...Main.args,
	error: undefined,
};

// export function UsingHooks(args) {
// 	const [, updateArgs] = useArgs();
//
// 	const props = useTForm<FormType>({
// 		...args,
// 		initialValues: {
// 			text: '',
// 			nested: {
// 				text: '',
// 			},
// 			plainInput: '',
// 		},
// 		validationSchema: formValidation,
// 		onSubmit: data => args.onSubmit(data),
// 	});
//
// 	return <MyForm {...props} />;
// }
//
// UsingHooks.args = {
// 	...Main.args,
// };
//
// export function OutsideSubmit(args) {
// 	const [, updateArgs] = useArgs();
// 	const formSubmitFn = useRef<() => Promise<void>>();
//
// 	return (
// 		<>
// 			<TForm<FormType>
// 				{...args}
// 				initialValues={{text: '', nested: {text: ''}, plainInput: ''}}
// 				validationSchema={formValidation}
// 				getSubmitFn={sub => (formSubmitFn.current = sub)}
// 				onSubmit={async (data, formikHelpers) => {
// 					args.onSubmit(data, formikHelpers);
// 				}}
// 			>
// 				{props => <MyForm {...props} />}
// 			</TForm>
// 			<Segment>
// 				<Button color="orange" onClick={() => formSubmitFn.current && formSubmitFn.current()}>
// 					Outside Submit
// 				</Button>
// 			</Segment>
// 		</>
// 	);
// }
//
// OutsideSubmit.args = {
// 	...Main.args,
// };
