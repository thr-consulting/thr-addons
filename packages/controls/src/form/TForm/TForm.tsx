import debug from 'debug';
import React from 'react';
import {Formik, FormikConfig, FormikValues} from 'formik';
import {TFormInner, TFormProps} from './TFormInner';

const d = debug('thx.controls.TForm');

export default function TForm<Values extends FormikValues = FormikValues>(
	props: FormikConfig<Values> & TFormProps<Values>,
) {
	return (
		<Formik<Values> {...props}>
			{formikProps => {
				return <TFormInner<Values> formikProps={formikProps} tFormProps={props} children={props.children} />;
			}}
		</Formik>
	);
}
