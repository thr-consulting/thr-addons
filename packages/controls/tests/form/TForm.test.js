/* eslint-disable react/prop-types */
import React from 'react';
import {mount} from 'enzyme';
import {Form} from 'semantic-ui-react';
import TForm from '../../src/form/TForm';

function renderForm({values, handleChange, handleBlur, handleSubmit, renderErrors, renderWarnings, hasErrors, hasWarnings, fieldError}) {
	return (
		<Form onSubmit={handleSubmit} error={hasErrors()} warning={hasWarnings()}>
			<Form.Field error={fieldError('text')}>
				<label>Text</label>
				<input
					name="text"
					value={values.text}
					onChange={handleChange}
					onBlur={handleBlur}
				/>
			</Form.Field>
			{renderErrors()}
			{renderWarnings()}
		</Form>
	);
}

function validate(values, props) {
	let warnings = {};
	//if (!values.text) {
		warnings.text = 'There is a problem with the text field.';
	//}
	return warnings;
}

function fakeSubmit() {
	return 'done!';
}

describe('TForm', () => {
	it('should render without throwing an error', () => {
		expect(mount(
			<TForm
				validate={validate}
				render={renderForm}
				initialValues={{text: 'Hello World!'}}
				loading={false}
				onSubmit={fakeSubmit}
				numFields={1}
			/>
		)).toMatchSnapshot();
	});

	// TODO This test doesn't seem to work after upgrading to formik 1.3.0 from 0.11.11
	// The actual component still works as desires though. See the Storybook control.
	// it('should render with warnings', () => {
	// 	const wrapper = mount(
	// 		<TForm
	// 			validate={validate}
	// 			render={renderForm}
	// 			loading={false}
	// 			onSubmit={fakeSubmit}
	// 			numFields={1}
	// 			initialValues={{text: ''}}
	// 		/>
	// 	);
	// 	const i = wrapper.find('input');
	// 	// console.log(wrapper.exists('input'), i.html());
	// 	i.simulate('focus', {target: {value: '', name: 'text'}});
	// 	// i.simulate('change', {target: {value: '', name: 'text'}})
	// 	i.simulate('blur', {target: {value: '', name: 'text'}});
	// 	expect(wrapper).toMatchSnapshot();
	// });

	it('should render with errors', () => {
		const wrapper = mount(
			<TForm
				validate={validate}
				render={renderForm}
				loading={false}
				onSubmit={fakeSubmit}
				numFields={1}
				errors={{graphQLErrors: [{message: 'There is an error.'}]}}
			/>
		);
		expect(wrapper).toMatchSnapshot();
	});
});
