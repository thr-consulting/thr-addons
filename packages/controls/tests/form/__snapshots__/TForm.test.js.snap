// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TForm should render with errors 1`] = `
<TForm
  errors={
    Object {
      "graphQLErrors": Array [
        Object {
          "message": "There is an error.",
        },
      ],
    }
  }
  loading={false}
  numFields={1}
  onSubmit={[Function]}
  render={[Function]}
  validate={[Function]}
>
  <Formik
    enableReinitialize={false}
    errors={
      Object {
        "graphQLErrors": Array [
          Object {
            "message": "There is an error.",
          },
        ],
      }
    }
    isInitialValid={false}
    loading={false}
    onSubmit={[Function]}
    render={[Function]}
    validate={[Function]}
    validateOnBlur={true}
    validateOnChange={true}
  >
    <Form
      as="form"
      error={true}
      onSubmit={[Function]}
      warning={false}
    >
      <form
        className="ui error form"
        onSubmit={[Function]}
      >
        <FormField
          error={false}
        >
          <div
            className="field"
          >
            <label>
              Text
            </label>
            <input
              name="text"
              onBlur={[Function]}
              onChange={[Function]}
            />
          </div>
        </FormField>
        <Message
          error={true}
        >
          <div
            className="ui error message"
          >
            <MessageHeader>
              <div
                className="header"
              >
                Error
              </div>
            </MessageHeader>
            <p
              key="There is an error."
            >
              There is an error.
            </p>
          </div>
        </Message>
      </form>
    </Form>
  </Formik>
</TForm>
`;

exports[`TForm should render without throwing an error 1`] = `
<TForm
  initialValues={
    Object {
      "text": "Hello World!",
    }
  }
  loading={false}
  numFields={1}
  onSubmit={[Function]}
  render={[Function]}
  validate={[Function]}
>
  <Formik
    enableReinitialize={false}
    initialValues={
      Object {
        "text": "Hello World!",
      }
    }
    isInitialValid={false}
    loading={false}
    onSubmit={[Function]}
    render={[Function]}
    validate={[Function]}
    validateOnBlur={true}
    validateOnChange={true}
  >
    <Form
      as="form"
      error={false}
      onSubmit={[Function]}
      warning={false}
    >
      <form
        className="ui form"
        onSubmit={[Function]}
      >
        <FormField
          error={false}
        >
          <div
            className="field"
          >
            <label>
              Text
            </label>
            <input
              name="text"
              onBlur={[Function]}
              onChange={[Function]}
              value="Hello World!"
            />
          </div>
        </FormField>
      </form>
    </Form>
  </Formik>
</TForm>
`;
